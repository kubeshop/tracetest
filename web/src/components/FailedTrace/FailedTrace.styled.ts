import {CloseCircleFilled} from '@ant-design/icons';
import styled from 'styled-components';

export const FailedTrace = styled.div`
  height: calc(100vh - 200px);
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

export const Container = styled.div`
  max-width: 520px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
`;

export const FailedIcon = styled(CloseCircleFilled).attrs({
  style: {
    color: '#e84749',
    fontSize: 32,
  },
})`
  margin-bottom: 26px;
`;

export const TextContainer = styled(Container)`
  margin-bottom: 26px;
`;

export const ButtonContainer = styled.div`
  display: flex;
  gap: 8px;
`;
