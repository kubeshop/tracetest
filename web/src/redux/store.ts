import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import TestAPI from 'redux/apis/Test.api';
import ResultListSlice from 'redux/slices/ResultList.slice';

export const store = configureStore({
  reducer: {
    [TestAPI.reducerPath]: TestAPI.reducer,
    resultList: ResultListSlice,
  },
  middleware: getDefaultMiddleware => getDefaultMiddleware().concat(TestAPI.middleware),
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
