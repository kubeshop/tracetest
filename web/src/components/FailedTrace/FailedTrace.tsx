import {Typography} from 'antd';
import {DISCORD_URL, GITHUB_ISSUES_URL} from 'constants/Common.constants';
import TestRun from 'models/TestRun.model';
import * as S from './FailedTrace.styled';

interface IProps {
  isDisplayingError: boolean;
  run: TestRun;
}

const FailedTrace = ({isDisplayingError, run: {lastErrorState}}: IProps) => {
  return isDisplayingError ? (
    <S.FailedTrace>
      <S.Container>
        <S.FailedIcon />
        <S.TextContainer>
          <Typography.Title level={1}>Test Run Failed</Typography.Title>
          <Typography.Text type="secondary">{lastErrorState}</Typography.Text>
          <Typography.Text type="secondary">
            Please let us know about this issue - <a href={GITHUB_ISSUES_URL}>create an issue</a> or contact us via{' '}
            <a href={DISCORD_URL}>Discord</a>.
          </Typography.Text>
          <Typography.Text type="secondary">We will check it out and respond to you.</Typography.Text>
        </S.TextContainer>
      </S.Container>
    </S.FailedTrace>
  ) : null;
};

export default FailedTrace;
