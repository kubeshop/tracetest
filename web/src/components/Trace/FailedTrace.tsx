import {Button, Typography} from 'antd';
import {useCallback} from 'react';
import {DISCORD_URL, GITHUB_ISSUES_URL} from '../../constants/Common.constants';
import {useRunTestMutation} from '../../redux/apis/Test.api';
import {ITestRunResult} from '../../types/TestRunResult.types';
import * as S from './FailedTrace.styled';

interface IFailedTraceProps {
  isDisplayingError: boolean;
  testId: string;
  onEdit(): void;
  onRunTest(result: ITestRunResult): void;
}

const FailedTrace: React.FC<IFailedTraceProps> = ({onRunTest, onEdit, testId, isDisplayingError}) => {
  const [runNewTest] = useRunTestMutation();

  const onReRun = useCallback(async () => {
    const result = await runNewTest(testId).unwrap();
    onRunTest(result);
  }, [onRunTest, runNewTest, testId]);

  return isDisplayingError ? (
    <S.FailedTrace>
      <S.Container>
        <S.FailedIcon />
        <S.TextContainer>
          <Typography.Title level={3}>Test Run Failed</Typography.Title>
          <Typography.Text type="secondary">Information explaining the state the test failed at.</Typography.Text>
          <Typography.Text type="secondary">
            Please let us know about this issue - <a href={GITHUB_ISSUES_URL}>create an issue</a> or contact us via{' '}
            <a href={DISCORD_URL}>Discord</a>.
          </Typography.Text>
          <Typography.Text type="secondary">We will check it out and respond to you.</Typography.Text>
        </S.TextContainer>
        <S.ButtonContainer>
          <Button type="primary" ghost onClick={onEdit}>
            Edit Test
          </Button>
          <Button type="primary" ghost onClick={onReRun}>
            Rerun Test
          </Button>
        </S.ButtonContainer>
      </S.Container>
    </S.FailedTrace>
  ) : null;
};

export default FailedTrace;
