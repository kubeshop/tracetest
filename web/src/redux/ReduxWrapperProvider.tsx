// test-utils.jsx
import * as React from 'react';
import {Provider} from 'react-redux';
import {store} from './store';

export const ReduxWrapperProvider: React.FC = ({children}) => {
  return <Provider store={store}>{children}</Provider>;
};
