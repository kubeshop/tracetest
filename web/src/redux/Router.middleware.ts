import {createListenerMiddleware} from '@reduxjs/toolkit';
import type {TypedStartListening} from '@reduxjs/toolkit';
import {LOCATION_CHANGE} from 'redux-first-history';
import {parse} from 'query-string';
import RouterActions from './actions/Router.actions';
import {runOriginPathAdded} from './slices/User.slice';
import {AppDispatch, RootState} from './store';

type AppStartListening = TypedStartListening<RootState, AppDispatch>;
const listener = createListenerMiddleware();
const startAppListening = listener.startListening as AppStartListening;

const sideEffectActionList = [RouterActions.updateSelectedAssertion, RouterActions.updateSelectedSpan];

const runUrlRegex = /^\/(test|testsuite)\/([^\/]+)\/run\/([^\/]+)(.*)$/;

const RouterMiddleware = () => ({
  middleware: listener.middleware,

  startListening(params = {}) {
    return startAppListening({
      predicate: ({type = ''}) =>
        type === LOCATION_CHANGE || (type.endsWith('/fulfilled') && !type.startsWith('router/')),
      effect(_, {dispatch, getState}) {
        const {
          router: {location},
        } = getState();

        const search = parse(location?.search || '');

        sideEffectActionList.forEach(action => {
          dispatch(action({search, params}));
        });
      },
    });
  },

  startListeningForLocationChange() {
    return startAppListening({
      predicate: action => {
        const pathname = action?.payload?.location?.pathname ?? '';
        return action?.type === LOCATION_CHANGE && pathname.match(runUrlRegex);
      },
      effect: async (_, {dispatch, getOriginalState, getState}) => {
        const {
          router: {location: prevLocation},
        } = getOriginalState();

        const {
          router: {location: currLocation},
        } = getState();

        const prevPathname = prevLocation?.pathname ?? '';

        if (!prevPathname.match(runUrlRegex) && !prevPathname.includes('/create')) {
          const defaultPath = currLocation?.pathname?.includes('testsuite') ? '/testsuites' : '/tests';
          dispatch(runOriginPathAdded(prevLocation?.pathname ?? defaultPath));
        }
      },
    });
  },
});

export default RouterMiddleware();
