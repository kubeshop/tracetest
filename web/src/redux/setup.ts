import {Middleware} from '@reduxjs/toolkit';

import OtelRepoAPI from 'redux/apis/OtelRepo';
import TestSpecs from 'redux/slices/TestSpecs.slice';
import Spans from 'redux/slices/Span.slice';
import DAG from 'redux/slices/DAG.slice';
import Trace from 'redux/slices/Trace.slice';
import CreateTestSuite from 'redux/slices/CreateTestSuite.slice';
import User from 'redux/slices/User.slice';
import TestOutputs from 'redux/testOutputs/slice';

export const middlewares: Middleware[] = [OtelRepoAPI.middleware];

export const reducers = {
  [OtelRepoAPI.reducerPath]: OtelRepoAPI.reducer,

  spans: Spans,
  dag: DAG,
  trace: Trace,
  testSpecs: TestSpecs,
  createTestSuite: CreateTestSuite,
  user: User,
  testOutputs: TestOutputs,
};
