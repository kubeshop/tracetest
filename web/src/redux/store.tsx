import {Action, configureStore, ThunkAction} from '@reduxjs/toolkit';
import {createReduxHistoryContext} from 'redux-first-history';
import {createBrowserHistory} from 'history';
import TracetestAPI from 'redux/apis/Tracetest';

import RouterMiddleware from './Router.middleware';
import {middlewares, reducers} from './setup';

const {createReduxHistory, routerMiddleware, routerReducer} = createReduxHistoryContext({
  history: createBrowserHistory(),
});

TracetestAPI.create();

export const store = configureStore({
  reducer: {
    ...reducers,
    [TracetestAPI.instance.reducerPath]: TracetestAPI.instance.reducer,
    router: routerReducer,
  },
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware()
      .prepend(RouterMiddleware.middleware)
      .concat(TracetestAPI.instance.middleware, ...middlewares, routerMiddleware),
});

export const history = createReduxHistory(store);

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
