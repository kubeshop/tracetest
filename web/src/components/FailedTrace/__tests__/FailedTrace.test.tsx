import {render} from '@testing-library/react';
import {MemoryRouter} from 'react-router-dom';
import FailedTrace from '../index';
import {ReduxWrapperProvider} from '../../../redux/ReduxWrapperProvider';

test('FailedTrace', () => {
  const result = render(
    <MemoryRouter>
      <ReduxWrapperProvider>
        <FailedTrace onRunTest={jest.fn()} testId="234" isDisplayingError onEdit={jest.fn()} />
      </ReduxWrapperProvider>
    </MemoryRouter>
  );
  expect(result.container).toMatchSnapshot();
});
