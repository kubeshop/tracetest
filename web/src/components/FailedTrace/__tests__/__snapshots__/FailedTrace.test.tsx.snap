// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FailedTrace 1`] = `
<div>
  <div
    class="sc-bdvvtL mLXGX"
  >
    <div
      class="sc-gsDKAQ hUQIyF"
    >
      <span
        aria-label="close-circle"
        class="anticon anticon-close-circle sc-dkPtRN ikUQFp"
        role="img"
        style="color: rgb(232, 71, 73); font-size: 32px;"
      >
        <svg
          aria-hidden="true"
          data-icon="close-circle"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
          />
        </svg>
      </span>
      <div
        class="sc-gsDKAQ sc-hKwDye hUQIyF ctlTiW"
      >
        <h3
          class="ant-typography"
        >
          Test Run Failed
        </h3>
        <span
          class="ant-typography ant-typography-secondary"
        >
          Information explaining the state the test failed at.
        </span>
        <span
          class="ant-typography ant-typography-secondary"
        >
          Please let us know about this issue - 
          <a
            href="https://github.com/kubeshop/tracetest/issues/new/choose"
          >
            create an issue
          </a>
           or contact us via
           
          <a
            href="https://discord.gg/6zupCZFQbe"
          >
            Discord
          </a>
          .
        </span>
        <span
          class="ant-typography ant-typography-secondary"
        >
          We will check it out and respond to you.
        </span>
      </div>
      <div
        class="sc-eCImPb hBTGvO"
      >
        <button
          class="ant-btn ant-btn-primary ant-btn-background-ghost"
          type="button"
        >
          <span>
            Edit Test
          </span>
        </button>
        <button
          class="ant-btn ant-btn-primary ant-btn-background-ghost"
          type="button"
        >
          <span>
            Rerun Test
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
